### PR TITLE
Add BuildContext.mounted

### LINKED ISSUE #111488: Expose a mounted-like flag on BuildContext
The popular `use_build_context_synchronously` lint suggests to check for `mounted` before using the `context` after an asynchronous gap. However, `mounted` only exists on `StatefullWidget`s. At the same time, `StatelessWidget`s suffer under the same problem: The `context` can become invalid after an asynchronous gap and there is currently no way to check whether it is still valid since there is no `mounted` property. A work-around is to turn the `StatelessWidget` into a `StatefulWidget`.

We should expose some kind of a property on `BuildContext` directly to check whether the context is still valid to simplify usage of `context` after an asynchronous gaps in these use cases.

### PR DESCRIPTION
**Pull request by goderbauer**
This makes it a little easier to use `BuildContext` error-free across asynchronous gaps in `StatelessWidget`s and other instances where you can't check `State.mounted`.

Fixes https://github.com/flutter/flutter/issues/111488

**DO NOT MERGE** until `use_build_context_synchronously` lint has been updated to recognize `BuildContext.mounted`: https://github.com/dart-lang/sdk/issues/58872

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/framework.dart:2091]: This is almost a functional example app. Would it make sense to make it into a full sample?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart:2091]: Hm, it doesn't demonstrate anything visually though. In this case I think a full sample would actually distract from the point this is trying to make...

Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/framework.dart:2091]: Ok, makes sense.  I mean it could show a dialog going away after a second, but I agree it's not all that visual.

**Review by chunhtai — APPROVED** [review]
LGTM

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by fzyzcjy** [comment]
Not an expert here, but I see `State.mounted` is sometimes not enough for a `State.context` to be used.

What I have seen: error, `Looking up a deactivated widget's ancestor is unsafe.`, when using `SomeInheritedWidget.of(context)` - you know, the very common practice. The error is like

```
Looking up a deactivated widget's ancestor is unsafe.
At this point the state of the widget's element tree is no longer stable.
To safely refer to a widget's ancestor in its dispose() method, save a reference to the ancestor by calling dependOnInheritedWidgetOfExactType() in the widget's didChangeDependencies() method.

When the exception was thrown, this was the stack: 
#0      Element._debugCheckStateIsActiveForAncestorLookup.<anonymous closure> (package:flutter/src/widgets/framework.dart:3990:9)
#1      Element._debugCheckStateIsActiveForAncestorLookup (package:flutter/src/widgets/framework.dart:4004:6)
#2      Element.dependOnInheritedWidgetOfExactType (package:flutter/src/widgets/framework.dart:4019:12)
#3      SomeOfMyInheritedWidget.of (...)
```

That seems to come from when the widget is deactivated and not yet (re) activated.

Notice it uses `_debugCheckStateIsActiveForAncestorLookup`, and your example code, `Navigator.of(context)`, calls `findRootAncestorStateOfType` which also has the `_debugCheckStateIsActiveForAncestorLookup` assertion. 
Therefore, I suspect this PR will suffer from the same error.

File: packages/flutter/lib/src/widgets/framework.dart
Comment (fzyzcjy) [inline packages/flutter/lib/src/widgets/framework.dart:2098]: (here)

**Review by fzyzcjy — COMMENTED** [review]
Not an expert here, but I see `State.mounted` is sometimes not enough for a `State.context` to be used.

What I have seen: error, `Looking up a deactivated widget's ancestor is unsafe.`, when using `SomeInheritedWidget.of(context)` - you know, the very common practice. The error is like

```
Looking up a deactivated widget's ancestor is unsafe.
At this point the state of the widget's element tree is no longer stable.
To safely refer to a widget's ancestor in its dispose() method, save a reference to the ancestor by calling dependOnInheritedWidgetOfExactType() in the widget's didChangeDependencies() method.

When the exception was thrown, this was the stack: 
#0      Element._debugCheckStateIsActiveForAncestorLookup.<anonymous closure> (package:flutter/src/widgets/framework.dart:3990:9)
#1      Element._debugCheckStateIsActiveForAncestorLookup (package:flutter/src/widgets/framework.dart:4004:6)
#2      Element.dependOnInheritedWidgetOfExactType (package:flutter/src/widgets/framework.dart:4019:12)
#3      SomeOfMyInheritedWidget.of (...)
```

That seems to come from when the widget is deactivated and not yet (re) activated.

My personal workaround is that, has an extra variable tracking activate/deactivate, and never call them when it is deactivated and not yet activated. But that does not sound elegant.

Notice it uses `_debugCheckStateIsActiveForAncestorLookup`, and your example code, `Navigator.of(context)`, calls `findRootAncestorStateOfType` which also has the `_debugCheckStateIsActiveForAncestorLookup` assertion. 
Therefore, I suspect this PR will suffer from the same error.

**Comment by goderbauer** [comment]
@fzyzcjy Can you file an issue for that? Ideally with some code that reproduces the error you're seeing? I am curious to see how you can do a dependency lookup while a global key move or reparenting is in progress.

**Comment by fzyzcjy** [comment]
@goderbauer Sure I will do that when see the bug again

**Comment by rhydiant** [comment]
@goderbauer - is it possible to see which Flutter release this will end up in? I'm getting `use_build_context_synchronously` lint warnings with Flutter `3.3.5` and the docs refer to `BuildContext.mounted` but it doesn't appear to be supported yet

**Comment by ashishbeck** [comment]
I can see this is currently available in the beta branch but not stable as of `3.3.6` Do we have a timeline for this in stable?

**Comment by ChangJoo-Park** [comment]
BuildContext.mounted does not appear in 3.3.5

**Comment by ValentinVignal** [comment]
> BuildContext.mounted does not appear in 3.3.5

@ChangJoo-Park I believe we will need to wait for a least the next minor version (>= 3.4.0)

**Comment by Troyx21** [comment]
It's been *TWO MONTHS* after this was merged and still linter warnings are present, linking to the [*invalid* docs page](https://dart-lang.github.io/linter/lints/use_build_context_synchronously.html). Why should this wait for the next minor release, shouldn't this have been the immediate patch (bugfix)? This should be considered bugfix, as linter page is a bug (it provides solution that doesn't build).

**Comment by gspencergoog** [comment]
@Troyx21 This feature is available in the beta and master channels.  Please take a look at the [description of our release channels ](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels) to decide if you want to switch to one of those channels, but we typically only release to the stable channel once a quarter or so (every three months), so it is expected for a change to take about that long to get to the stable channel from the time it is committed to master. If you absolutely need the feature sooner than that, you are welcome to switch to a less stable channel in the meantime.

**Comment by Troyx21** [comment]
@gspencergoog I do understand that perfectly well. What I tried to communicate, is that this particular issue, because of the **broken** linter page and annoing linter warning that affects probably thouthands of codebases, could have been identified as a **bug**, therefore - this PR could have been labeled as a **bugfix** - and hence merged into stable as soon as the next bugfix (patch) version was issued (3.3.**X**).

**Comment by gspencergoog** [comment]
I'm sorry that it's annoying to you, I truly am.  However, we are very unlikely to elevate any annoyance issue to the level of a cherrypick/hotfix into the stable channel because the risk to the stability of the stable channel is too high. We [set the bar very high](https://github.com/flutter/flutter/wiki/Flutter-Cherrypick-Process) for those hotfixes because we want you to have confidence in the stability of the stable channel, as most developers would rather have annoying linter messages than angry customers. As [this page](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#will-a-particular-bug-fix-be-provided-in-a-hotfix-release) mentions, you are also welcome to patch your local version of Flutter if you must be on the stable channel but need the fix sooner.

**Comment by guplem** [comment]
Is context.mounted already available in version 3.3.10? I am getting the error `The getter 'mounted' isn't defined for the type 'BuildContext'. `

This is my flutter doctor output:
[√] Flutter (Channel stable, 3.3.10, on Microsoft Windows [Version 10.0.22621.819], locale en-150)
[√] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Build Tools 2019 16.11.19)
[√] Android Studio (version 2021.3)
[√] VS Code (version 1.74.1)
[√] Connected device (3 available)
[√] HTTP Host Availability

• No issues found!

This is my code:

```
loadConfigurationFromLocalStorage(BuildContext context) async {
  final localStorage = await SharedPreferences.getInstance();

  if (!context.mounted) {  // <-- Here I get the error 
    print("Context not mounted while trying to load configuration from local storage");
    return;
  }

  final String? storedLanguage = localStorage.getString("lang");
  if (storedLanguage != null) {
    Provider.of<AppProvider>(context, listen: false).setLanguage(storedLanguage);
  }

}
```

It might be the case that I am using it wrong, but if this is not a proper way of loading data from the local storage and saving it to a provider, how should I do it?

**Comment by ValentinVignal** [comment]
@guplem I don't belong to the flutter team so everything I'm saying is not confirmed. It's only suppositions.

> Is context.mounted already available in version 3.3.10? I am getting the error The getter 'mounted' isn't defined for the type 'BuildContext'.

Since `context.mounted` is a new feature, I doubt they will release it in a patch version (`3.3.x`), those are reserved for hotfixes. It will be at least a new minor version (`>= 3.4.0`) (as I said in my comment https://github.com/flutter/flutter/pull/111619#issuecomment-1306718556) You can read more about [semantic versioning](https://semver.org/).

>  3 months after the merge "seem enough" to release it into the stable version…

Flutter releases a new stable version _"roughly once a quarter"_ , so a new stable version should be released soon (but I don't know when). You can read more about the [Flutter build release channels](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels).
But there is also no guarantee `context.mounted` will be included in the next stable release (ex: `3.7.0`).

**Comment by rupinderjeet** [comment]
Until then, can something like this work?

```dart
extension BuildContextMountedExt on BuildContext {
  bool get mounted {
    try {
      widget;
      return true;
    } catch (e) {
      return false;
    }
  }
}
```

**Comment by ValentinVignal** [comment]
It looks like it has been released with Flutter 3.7.0 🎉

**Pull request merged**
